packages/logging/src/analytics-helpers.ts

```diff
@@ -378,3 +378,72 @@ export class ThrottledAnalytics implements MongoshAnalytics {
     });
   }
 }
+
+type SampledAnalyticsOptions = {
+  target?: MongoshAnalytics;
+  /**
+   * Sampling options. If not provided, sampling will be defaulted to 30% of sessions.
+   * Also, from an exposed environment standpoint, providing a MONGOSH_ANALYTICS_SAMPLE
+   * environment variable with a truthy value will force the sampling to 100%.
+   *
+   * The sampling configuration can be changed, however, by default it will be a Math.random
+   * bounded from 0 to 100. If you are going to configure it use a evenly distributed random function.
+   */
+  sampling: {
+    percentage: number;
+    samplingFunction: (percentage: number) => boolean;
+  } | null;
+};
+
+export class SampledAnalytics implements MongoshAnalytics {
+  private isEnabled: boolean;
+  private target: MongoshAnalytics;
+
+  private constructor(configuration: SampledAnalyticsOptions) {
+    configuration.sampling ??= {
+      percentage: 30,
+      samplingFunction: (percentage) => Math.random() * 100 < percentage,
+    };
+
+    const shouldBeSampled = configuration.sampling.samplingFunction(
+      configuration.sampling.percentage
+    );
+
+    this.isEnabled = !!process.env.MONGOSH_ANALYTICS_SAMPLE || shouldBeSampled;
```

Comment:
I think it might be a bit surprising that this is where we check the environment variable value – generally speaking, we try to keep as much interaction with the process state in `cli-repl.ts` and `run.ts` in the cli-repl package, and in the other parts of the codebase just rely on explicitly passing information

I’d still think that it’s fair to simplify the interface here to taking a `boolean` or maybe `() => boolean`, passing something like `() => !!process.env.MONGOSH_ANALYTICS_SAMPLE || Math.random() < 0.3` from the cli-repl package doesn’t seem too complex or untested to me